src/Language/PureScript/Scheme/CodeGen/Optimizer.hs

```diff
@@ -17,10 +17,21 @@ optimizations = (simplifyLogic . inlineCommonBinaryOperators)
 inlineCommonBinaryOperators :: SExpr -> SExpr
 
 inlineCommonBinaryOperators (List [List [List [Symbol op, Symbol klass], x], y])
-  | op == "Data.Semiring.add" && klass == "Data.Semiring.semiringInt" =
-    app "scm:+" [x, y]
-  | op == "Data.Ring.sub" && klass == "Data.Ring.ringInt" =
-    app "scm:-" [x, y]
+  | klass == "Data.Semiring.semiringInt" && op == "Data.Semiring.add"
+  = app "scm:+" [x, y]
+  | klass == "Data.Ring.ringInt" && op == "Data.Ring.sub"
+  = app "scm:-" [x, y]
+  | klass == "Data.Ring.ringNumber" && op == "Data.Ring.sub"
```

Comment:
Since we can distinguish between different types (Int, Number, ...) we can use specialized arithmetic functions ([fx+](https://scheme.com/tspl4/objects.html#./objects:s157) vs. [fl+](https://scheme.com/tspl4/objects.html#./objects:s184), [fx*](https://scheme.com/tspl4/objects.html#./objects:s159) vs. [fl*](https://scheme.com/tspl4/objects.html#./objects:s186), ...).

Or at least leave a TODO here.

Reply:
Let's open a ticket to track this